### PR TITLE
Update major interfaces

### PR DESCRIPTION
**Pull request by junha1**
1. Use anyhow::Error
2. Trait -> Struct: this is because we don't want to be generic about
the impl, but the combination of the underlying primitives (network and
storage)
3. Put `Send + Sync + 'static` by default

### CODE REVIEW

File: network/src/dms.rs [hunk 1]
```diff
@@ -1,65 +1,151 @@
...
+    #[allow(dead_code)]
+    pub fn into_message(self) -> anyhow::Result<Message> {
+        Message::new(self.data, self.signature).map_err(|e| anyhow!(e))
+    }
+
+    #[allow(dead_code)]
+    pub fn from_message(message: Message) -> Self {
+        RawMessage {
+            data: message.data().to_owned(),
+            signature: message.signature().clone(),
```
Comment (lemonholic) [inline network/src/dms.rs:54]: Why not `to_owned()`?

File: network/src/dms.rs [hunk 2]
```diff
@@ -1,65 +1,151 @@
...
+        }
+    }
+}
 
 /// The interface that will be wrapped into an HTTP RPC server for the peers.
-#[async_trait]
-pub trait DistributedMessageSetRpcInterface {
+#[serde_tc_full]
+trait DistributedMessageSetRpcInterface: Send + Sync + 'static {
     /// Returns the messages except `knowns`. If the height is different, it returns `Err(height)`.
```
Comment (lemonholic) [inline network/src/dms.rs:62]: 1. It seems `DmsRpcInterface::get_message` (at line 63) doesn't return `Err(height)` anymore.
2. Why did you change the error type?

File: network/src/dms.rs [hunk 3]
```diff
@@ -1,65 +1,151 @@
...
+    /// - `dms_key`: The unique key for distinguishing the DMS.
+    pub async fn create(_storage: S, _height: u64, _dms_key: String) -> Result<(), Error> {
         unimplemented!()
     }
 
-    async fn open(_storage_directory: &str) -> Result<Self, Error>
+    /// Opens an existing storage with the given directory.
+    ///
+    /// - `dms_key`: The unique key for distinguishing the DMS.
+    pub async fn open(_storage: S) -> Result<Self, Error>
```
Comment (lemonholic) [inline network/src/dms.rs:100]: but where's the key?

File: network/src/dms.rs (hunk 2 above)
Comment (junha1) [reply network/src/dms.rs:62]: 1. Fixed.
2. I found that there could be more error cases besides height mismatch.

File: network/src/dms.rs (hunk 3 above)
Comment (junha1) [reply network/src/dms.rs:100]: Not needed. Removed.

File: network/src/dms.rs (hunk 1 above)
Comment (junha1) [reply network/src/dms.rs:54]: both `clone()` and `to_owned()` perform exactly the same, but `to_owned()` means more proper, afaik.
I will update it, but note that I didn't really use `to_owned()` in other code.

File: network/src/dms.rs (hunk 2 above)
Comment (lemonholic) [reply network/src/dms.rs]: alright

File: network/src/dms.rs (hunk 1 above)
Comment (lemonholic) [reply network/src/dms.rs]: checked

File: network/src/primitives.rs [hunk 4]
```diff
@@ -2,20 +2,51 @@ use super::*;
 use async_trait::async_trait;
 use tokio::sync::mpsc;
 
+pub type StroageError = std::io::Error;
```
Comment (lemonholic) [inline network/src/primitives.rs:5]: `StroageError` 😱

File: network/src/primitives.rs [hunk 5]
```diff
@@ -2,20 +2,51 @@ use super::*;
...
+
+    /// Add the given file to the storage.
+    async fn add_or_overwrite_file(
+        &mut self,
+        path: String,
+        content: String,
+    ) -> Result<(), StroageError>;
+
+    /// Reads the given file.
+    async fn read_file(&self, path: &str) -> Result<String, StroageError>;
```
Comment (lemonholic) [inline network/src/primitives.rs:30]: Some paths are typed as `&str` while some are typed as `String`. (e.g. L20-27)

File: network/src/dms.rs (hunk 3 above)
Comment (lemonholic) [reply network/src/dms.rs:98]: ok

File: repository/src/lib.rs [hunk 6]
```diff
@@ -115,39 +110,134 @@ pub trait DistributedRepository<T: RawRepository>: Send + Sync {
...
+
+        let agenda_commit = Commit::Agenda(Agenda {
+            author,
+            timestamp: get_timestamp(),
+            hash: Agenda::calculate_hash(last_header.height + 1, &transactions),
+        });
+        let semantic_commit = to_semantic_commit(&agenda_commit, &last_header);
+
+        self.raw.checkout_clean().await?;
+        self.raw.checkout(&FINALIZED_BRANCH_NAME.into()).await?;
```
Comment (lemonholic) [inline repository/src/lib.rs:226]: Why do you checkout to `main` just before creating an agenda commit?

File: governance/src/lib.rs [hunk 7]
```diff
@@ -36,37 +16,73 @@ pub struct GovernanceState {
...
+            agenda_hash,
+            voter: private_key.public_key(),
+            signature: Signature::sign(agenda_hash, private_key)?,
+        })
+        .unwrap();
+        let message = Message::new(
+            data.clone(),
+            TypedSignature::sign(&data, &network_config.private_key)?,
+        )?;
+
```
Comment (TomTaehoonKim) [inline governance/src/lib.rs:60]: delete empty line?

**Comment by TomTaehoonKim** [comment]
typo in commit message of commit 3c0f58a

File: network/src/dms.rs [hunk 8]
```diff
@@ -1,65 +1,149 @@
...
 
-pub struct DistributedMessageSetImpl {}
+/// A **cumulative** set that is shared in the p2p network, backed by the local file system.
+///
+/// One of the notable characteristics of blockchain is that it is based on heights;
+/// The key idea here is that we retain an instance (both in memory or on disk)
+/// of `DistributedMessageSet` only for a specific height,
+/// and discard if the height progresses, creating a new and empty one again.
+///
+/// For every methods,
```
Comment (TomTaehoonKim) [inline network/src/dms.rs:77]:
```suggestion
/// For every method,
```

File: network/src/dms.rs [hunk 9]
```diff
@@ -1,65 +1,151 @@
...
+    _marker: std::marker::PhantomData<N>,
+}
 
-#[async_trait]
-impl DistributedMessageSet for DistributedMessageSetImpl {
-    async fn create(_storage_directory: &str, _height: u64) -> Result<(), Error> {
+impl<N: P2PNetwork, S: Storage> DistributedMessageSet<N, S> {
+    /// Creates a new and empty storage with the given directory.
+    /// If there is already a directory, it discards everything and creates a new one.
+    /// You should try `open()` first!
```
Comment (TomTaehoonKim) [inline network/src/dms.rs:90]: Anyway to improve this in a way s.t. opening an existing directory as default?

File: network/src/dms.rs [hunk 10]
```diff
@@ -1,65 +1,149 @@
...
+    pub async fn fetch(
         &mut self,
         _network_config: &NetworkConfig,
         _known_peers: &[Peer],
     ) -> Result<(), Error> {
         unimplemented!()
     }
 
-    async fn add_message(
+    /// Add the given message to the storage, immediately broadcasting it to the network.
```
Comment (TomTaehoonKim) [inline network/src/dms.rs:114]: Add -> Adds

File: governance/src/lib.rs (hunk 7 above)
Comment (junha1) [reply governance/src/lib.rs:60]: maybe, but not important. I'll skip this.

File: network/src/primitives.rs [hunk 11]
```diff
@@ -2,20 +2,51 @@ use super::*;
...
+
+    /// Opens an existing directory.
+    async fn open(storage_directory: &str) -> Result<Self, StroageError>
+    where
+        Self: Sized;
+
+    /// Shows the list of files.
+    async fn list_files(&self) -> Result<Vec<String>, StroageError>;
+
+    /// Add the given file to the storage.
```
Comment (TomTaehoonKim) [inline network/src/primitives.rs:22]: Add -> Adds

File: repository/src/lib.rs [hunk 12]
```diff
@@ -27,84 +28,78 @@ impl fmt::Display for CommitHash {
...
+    let since_the_epoch = now.duration_since(std::time::UNIX_EPOCH).unwrap();
+    since_the_epoch.as_millis() as Timestamp
+}
 
+impl<T: RawRepository> DistributedRepository<T> {
+    pub async fn new(_raw: T) -> Result<Self, Error> {
+        unimplemented!()
+    }
+
+    /// Initialize the genesis repository from the genesis working tree.
```
Comment (TomTaehoonKim) [inline repository/src/lib.rs:54]: Initialize -> Initializes

File: network/src/primitives.rs (hunk 4 above)
Comment (junha1) [reply network/src/primitives.rs]: fixed

File: network/src/primitives.rs (hunk 5 above)
Comment (junha1) [reply network/src/primitives.rs]: fixed

File: repository/src/lib.rs (hunk 6 above)
Comment (junha1) [reply repository/src/lib.rs]: fixed

File: network/src/dms.rs (hunk 8 above)
Comment (junha1) [reply network/src/dms.rs:77]: this comment will be replaced in total in the next PR. Let's keep it for now.

File: network/src/dms.rs (hunk 9 above)
Comment (junha1) [reply network/src/dms.rs:90]: that's not the role of this interface, but the role of the user.

File: repository/src/lib.rs (hunk 12 above)
Comment (junha1) [reply repository/src/lib.rs]: fixed

File: network/src/primitives.rs (hunk 11 above)
Comment (junha1) [reply network/src/primitives.rs]: fixed

File: network/src/dms.rs (hunk 10 above)
Comment (junha1) [reply network/src/dms.rs]: fixed

**Review by lemonholic — APPROVED** [review]
lgtm

**Review by TomTaehoonKim — APPROVED** [review]
LGTM
